resources/idlharness-shadowrealm.js

```diff
@@ -0,0 +1,78 @@
+// TODO: it would be nice to support `idl_array.add_objects`
+/**
+ * idl_test_shadowrealm is a promise_test wrapper that handles the fetching of the IDL, and
+ * running the code in a `ShadowRealm`, avoiding repetitive boilerplate.
+ *
+ * @see https://github.com/tc39/proposal-shadowrealm
+ * @param {String[]} srcs Spec name(s) for source idl files (fetched from
+ *      /interfaces/{name}.idl).
+ * @param {String[]} deps Spec name(s) for dependency idl files (fetched
+ *      from /interfaces/{name}.idl). Order is important - dependencies from
+ *      each source will only be included if they're already know to be a
+ *      dependency (i.e. have already been seen).
+ */
+function idl_test_shadowrealm(srcs, deps) {
+    const script_urls = [
+        "/resources/testharness.js",
+        "/resources/WebIDLParser.js",
+        "/resources/idlharness.js",
+    ];
+    promise_test(async t => {
+        const realm = new ShadowRealm();
+        realm.evaluate("globalThis.self = globalThis; undefined");
+
+        const ss = await Promise.all(script_urls.map(url => fetch_text(url)));
+        for (const s of ss) {
+            realm.evaluate(s);
+        }
+        const specs = await Promise.all(srcs.concat(deps).map(fetch_spec));
+        const idls = JSON.stringify(specs.map(i => i.idl));
+        const code = `
+            const idls = ${idls};
+            let results;
+            add_completion_callback(function (tests, harness_status, asserts_run) {
+                results = tests;
+            });
+
+            // Without the wrapping test, testharness.js will think it's done after it has run
```

Comment:
I guess this is because it's not in a window context, so it's similar to workers needing an explicit `done()`? So can we use that instead and avoid the wrapping test?

Reply:
I have looked briefly, but this part of th.js is... interesting. I'd be willing (though not very excited) to look into this more, but I'd prefer to land first so I can unblock my spec work.

Reply:
I feel like this is adding another layer of dependencies on weird behavior or testharness.js that none of us really understand, but I also can't be bothered to look into this now, so I can't block you :)